[feat] Adding UPerNet

brianhou0208 opened the pull request:
## Description 
I tried to add UperNet before. This may not be the best version. Maybe you can have a look.
@GrantorShadow @jawi289o
## Issue
#906 
## Reference
- [official repo](https://github.com/CSAILVision/unifiedparsing/blob/master/models/models.py)
- [mmseg repo](https://github.com/open-mmlab/mmsegmentation/blob/v0.17.0/mmseg/models/decode_heads/uper_head.py#L13)

qubvel commented:
Hi, @brianhou0208, thanks a lot for working on this! Looks good to me, a few general notes/questions
 - Can you please use more verbose variable naming, e.g. `ch->channels`, `h -> height` (it is not hold over the repo, but would be nice to have it)
 - Please, add the model to the docs/ and README.md
 - Did you have a chance to train it with any notebook example?

qubvel reviewed (commented):
Hi @brianhou0208, thanks a lot for working on this! Looks good to me, a few general notes/questions
 - Can you please use more verbose variable naming, e.g. `ch->channels`, `h -> height` (it is not hold over the repo, but would be nice to have it)
 - Please, add the model to the docs/ and README.md
 - Did you have a chance to train it with any notebook example?

brianhou0208 commented:
Hi @qubvel , I’ve updated the code according to the commit, along with the README and documentation.
I’ve also trained the model using the `binary_segmentation_intro.ipynb`, and the results are shown below.

Due to device limitations, the default parameters caused an OOM error, so I adjusted the `decoder_pyramid_channels` and `decoder_segmentation_channels` parameters.
- Valid Dataset

| Arch          | Backbone | pyramid_channels | segmentation_channels | IoU (Per Image / Dataset) |
| :------------ | :------: | :--------------: | :-------------------: | :-----------------------: |
| UPerNet       | ResNet34 |        32        |          32           |       90.51 / 91.35       |
| UPerNet       | ResNet34 |        32        |          16           |       90.48 / 91.38       |
| UPerNet       | ResNet34 |        32        |           8           |       90.01 / 90.92       |
| FPN           | ResNet34 |        -         |           -           |       90.56 / 91.39       |
| DeepLabV3Plus | ResNet34 |        -         |           -           |       90.26 / 91.07       |
| UNet          | ResNet34 |        -         |           -           |       90.50 / 91.31       |
- Test Dataset

| Arch          | Backbone | pyramid_channels | segmentation_channels | IoU (Per Image / Dataset) |
| :------------ | :------: | :--------------: | :-------------------: | :-----------------------: |
| UPerNet       | ResNet34 |        32        |          32           |       91.11 / 91.75       |
| UPerNet       | ResNet34 |        32        |          16           |       91.10 / 91.76       |
| UPerNet       | ResNet34 |        32        |           8           |       91.01 / 91.73       |
| FPN           | ResNet34 |        -         |           -           |       90.97 / 91.64       |
| DeepLabV3Plus | ResNet34 |        -         |           -           |       90.48 / 91.22       |
| UNet          | ResNet34 |        -         |           -           |       91.15 / 91.76       |

qubvel commented on segmentation_models_pytorch/decoders/upernet/decoder.py:
```diff
@@ -0,0 +1,138 @@
+
+        resized_fpn_features = []
+        for feature in fpn_features:
+            resized_feature = F.interpolate(
```
As far as I see from the original arch features are upsampled to 1/4 size of the original image, so the output feature maps has 4 times less resolution then the original one. We can include upsampling layer later similar to FPN architecture to make sure we return mask of the same shape as image

brianhou0208 replied:
I missed this detail. After checking, the 'fpn' output shouldn’t be resized to the original resolution. I've updated the code to fix it.

qubvel reviewed (commented):
Thanks for update, I made a quick look at the architecture design, here is main concern

qubvel reviewed (approved):
Looks great, just one nit and we are ready to merge this!

brianhou0208 commented:
I've finished it and fixed a few small issues.

qubvel commented:
@brianhou0208 Thanks a lot for addressing it!
